libraries/psibase/common/include/psibase/serveGraphQL.hpp

```diff
@@ -349,6 +349,120 @@ namespace psibase
       return result;
    }  // makeConnection
 
+   template <typename T, typename Key>
+   concept HasGetKey = requires(T t) {
+      {
+         t.getKey()
+      } -> std::same_as<Key>;
+   };
+
+   /// Similar to makeConnection, except that it allows pagination through a virtual table index.
+   /// A virtual table index is a dynamically constructed vector of objects.
+   template <typename Connection, typename T, typename Key>
+      requires HasGetKey<T, Key>
+   Connection makeVirtualConnection(const std::vector<T>&             elements,
+                                    const std::optional<Key>&         gt,
+                                    const std::optional<Key>&         ge,
+                                    const std::optional<Key>&         lt,
+                                    const std::optional<Key>&         le,
+                                    std::optional<uint32_t>           first,
+                                    std::optional<uint32_t>           last,
+                                    const std::optional<std::string>& before,
+                                    const std::optional<std::string>& after)
+   {
+      auto indexFromString = [&](const std::string& str, size_t& index) -> bool
+      {
+         char*         end;
+         unsigned long result = std::strtoul(str.c_str(), &end, 10);
+         if (end == str.c_str() || *end != '\0' || result == std::numeric_limits<size_t>::max())
+         {
+            return false;
+         }
+         index = static_cast<size_t>(result);
+         return true;
+      };
+
+      auto lower_bound = [&](const Key& key)
```

Comment:
If you pass in a proj function (`T -> Key`), instead of relying on `getKey`, then `std::ranges::lower_bound(elements, key, {}, proj)` will work.

Reply:
I like this. There's probably a more clever way to do it, but I just used std::function. That workable?

Reply:
Yes. Or make it a template parameter and `requires std::is_invokable_r<Key, F, const T&>`

Reply:
If you're okay with std function, let's proceed as is, otherwise lmk and I'll work on this